server/datastore/mysql/operating_systems.go

```diff
@@ -139,23 +139,12 @@ func getOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostOS fleet.
 // upsertHostOperatingSystemDB upserts the host operating system table
 // with the operating system id for the given host ID
 func upsertHostOperatingSystemDB(ctx context.Context, tx sqlx.ExtContext, hostID uint, osID uint) error {
-	res, err := tx.ExecContext(ctx, "UPDATE host_operating_system SET os_id = ? WHERE host_id = ?", osID, hostID)
-	if err != nil {
-		return err
-	}
-
-	if n, _ := res.RowsAffected(); n > 0 {
-		// update success
-		return nil
-	}
-
-	// no row to update so insert new row
-	_, err = tx.ExecContext(ctx, "INSERT INTO host_operating_system (host_id, os_id) VALUES (?, ?)", hostID, osID)
-	if err != nil {
-		return err
-	}
-
-	return nil
+	_, err := tx.ExecContext(
```

Comment:
The prior pattern follows the general approach of optimistic updates that is used throughout the backend (e.g., [here](https://github.com/fleetdm/fleet/blob/main/server/datastore/mysql/mysql.go#L1260 and [here]https://github.com/fleetdm/fleet/blob/main/server/datastore/mysql/hosts.go#L3287). Is there something in this particular case that is leading us to reverse the general approach?

Reply:
Using UPDATE followed by INSERT inside a transaction causes deadlocks.

In this case, UPDATE that does not affect any rows locks the entire table.

If 2 threads simultaneously do UPDATE that does not affect any rows, both of them lock the entire table, and neither of them can do the subsequent INSERT.

What is the downside for updating this specific case? This table does not have an AUTO_INCREMENT column, so we don't need to worry about incrementing the counter for every INSERT.

Reply:
The downside, as I understand it (based primarily on [this]https://fleetdm.com/handbook/engineering/scaling-fleet#insert-on-duplicate-update) explanation of the general approach to optimistic updates, for which Lucas or Martin may be able to offer additional historical context), is that more often than not this particular method will be used in cases where the row does already exist because it will be called every time an existing host reports detail query results. So while this would mitigate some deadlocks in the initial host enrollment scenario, it might in turn harm performance in the more common detail query reporting for existing host reports. 

Does this need to be done inside a transaction? I notice at least in some of the other spots where optimistic updates are currently being used that those are not wrapped in transactions.

Reply:
Yes, not having a transaction would prevent a deadlock. But transactions are typically a good idea. In this case, it would avoid the cleanup job messing with table rows.

It seems like we need two code paths, but that may make the initial enrollment more expensive.

Reply:
>  In this case, it would avoid the cleanup job messing with table rows.

Are you referring to this [cleanup](https://github.com/fleetdm/fleet/blob/main/server/datastore/mysql/operating_systems.go#L184) or something else?

Reply:
I updated `UpdateHostOperatingSystem` to add the path for the common case.